subsys/usb/device_next/class/usbd_cdc_ecm.c

```diff
@@ -510,14 +542,21 @@ static int cdc_ecm_send(const struct device *dev, struct net_pkt *const pkt)
 		return -ENOMEM;
 	}
 
-	if (net_pkt_read(pkt, buf->data, len)) {
-		LOG_ERR("Failed copy net_pkt");
-		net_buf_unref(buf);
+	tmp = pkt->frags;
+	while (tmp) {
+		size_t copied;
 
-		return -ENOBUFS;
-	}
+		copied = net_buf_append_bytes(buf, tmp->len, tmp->data,
```

Comment:
This can result in net buf with multiple fragments, right? Does the rest of the USB stack handler this?

Reply:
> This can result in net buf with multiple fragments, right? Does the rest of the USB stack handler this?

Nope, and how the frags might be organized in net_pkt is not how the USB host would expect the data. If net_buf_frags_len() can be larger than NET_ETH_MAX_FRAME_SIZE then we would need to allocate and enqueue another buffer.

> btw. To me the use of net_buf in USB stack sounds like an abuse. It really should be abstracted as urb (even if the implementation underneath is net_buf).

No, we do not abuse it, it is a general purpose buffer subsystem. However, wrapping it in a different structure and not using net_buf's user data might simplify some things, or not.

Reply:
Hmm, do you mean that USB subsys is not able to handle multiple net_buf fragments? It seems to work fine (after this PR) when I set the net_buf data size to 128 which is typical default when networking is enabled.
If USB requires that net_buf data size is set to 1500 in case of Ethernet, then it is fundamentally flawed and cannot work with networking subsystem.

Reply:
On top of that, UDC also requires that the net_bufs passed to it are aligned with regards to https://github.com/zephyrproject-rtos/zephyr/blob/66ff30efb4d786ac15a4ef03a81bbe7a3f315690/include/zephyr/drivers/usb/udc_buf.h#L44-L48
This is why there are macros like https://github.com/zephyrproject-rtos/zephyr/blob/66ff30efb4d786ac15a4ef03a81bbe7a3f315690/include/zephyr/drivers/usb/udc_buf.h#L119-L133

I think hiding this behind URB-type may be the simplest and cleanest way to enforce these requirements at build-time.

Reply:
Ok, understood. I will rework this PR a bit then, note that the network side net_buf's are typically 128 bytes long and must be handled with care in this respect.

Reply:
My bad, this PR is totally useless, lets ignore it.